File: java/bin2file/bin2file.java
//
// Put binary portion from file to file
// Copyright (C) 2021 Dmitry Stefankov
//

import java.io.*;
import java.util.List;
import java.nio.charset.Charset;
import java.nio.file.*;

public class bin2file {

  public static byte[] MyReadAllBytes(String filename) throws IOException {
	Path file = Paths.get(filename);
	return Files.readAllBytes(file);
  }

  public static void main(String[] args) throws IOException {
     //System.out.println("There are " + args.length + " arguments given.");
     //for(int i = 0; i < args.length; i++) 
    //    System.out.println("The argument #" + (i+1) + " is " + args[i] + " and is at index " + i);
     if (args.length < 5) {
       System.out.println("Put binary portion from file to file, version 1.0");
       System.out.println("Copyright (c) 2021, Dmitry Stefankov");
       System.out.println("Usage: bin2file.exe infile outfile offset size newoffset");
       return;
     }
 
     String InFile = args[0];
     String OutFile = args[1];
     System.out.println("InFile: "+InFile);
     System.out.println("OutFile: "+OutFile);
     int in_offset = 0, in_length = 0, out_offset = 0;
     try {
       in_offset = Integer.parseInt(args[2]);
       System.out.println("in_Offset: "+in_offset);
       in_length = Integer.parseInt(args[3]);
       System.out.println("in_Length: "+in_length);
       out_offset = Integer.parseInt(args[4]);
       System.out.println("out_Offset: "+out_offset);
     }
     catch (NumberFormatException ex){
        ex.printStackTrace();
     }

     byte[] ReadBuffer = MyReadAllBytes(InFile);
     int InBufSize = ReadBuffer.length;
     System.out.println("InBufSize: "+InBufSize);

     byte[] WriteBuffer = MyReadAllBytes(OutFile);
     int OutBufSize = WriteBuffer.length;
     System.out.println("OutBufSize: "+OutBufSize);

     for (int i = 0; i < in_length; i++) {
        WriteBuffer[i+out_offset] = ReadBuffer[i+in_offset];
     }

     try (FileOutputStream stream = new FileOutputStream(OutFile)) {
       stream.write(WriteBuffer);
     }

  }

}

File: haskell/bin2file/bin2file.hs
--
-- Put binary portion from file, version 1.0
-- Copyright (C) 2025 Dmitry Stefankov. All Rights Reserved.
--

module Main (main) where

import Control.Monad
import Control.Monad.Except
import System.Console.GetOpt
import System.Environment(getArgs, getProgName)
import System.IO
import System.Exit
import Data.Char (isSpace)
import Data.Word
import Data.Binary.Get
import Data.Binary.Put
import Data.ByteString(readFile)
import qualified Data.ByteString as B
import Data.Char (isSpace)

data Options = Options {
    optVerbose :: Bool
  , optShowVersion :: Bool
  , optInputFilename :: String
  , optOutputFilename :: String
  , optInOffset :: Int
  , optInBytes :: Int
  , optOutOffset :: Int
  } deriving Show

defaultOptions = Options {
    optShowVersion = False 
  , optVerbose = False
  , optInputFilename = ""
  , optOutputFilename = ""
  , optInOffset = 0
  , optInBytes = 0
  , optOutOffset = 0
  }

options :: [OptDescr (Options -> Either String Options)]
options =
  [ Option ['i'] ["infile"]
        (ReqArg (\arg opts -> Right opts { optInputFilename = return arg !! 0 } -- really dirty, sorry but could not make it work another way
        ) "INFILE")
        "input filename"
  , Option ['o'] ["outfile"]
        (ReqArg (\arg opts -> Right opts { optOutputFilename = return arg !! 0 } -- really dirty, sorry but could not make it work another way
        ) "OUTFILE")
        "output filename"
  , Option ['l'] ["inoffset"]
      (ReqArg (\l opts ->
        case reads l of
          [(inoffset, "")] | inoffset >= 0 -> Right opts { optInOffset = inoffset }
          _ -> Left "--input offset value must be a number zero or greater"
        ) "OFFSET")
      "byte offset on input stream"
  , Option ['s'] ["inbytes"]
      (ReqArg (\s opts ->
        case reads s of
          [(inbytes, "")] | inbytes >= 0 -> Right opts { optInBytes = inbytes }
          _ -> Left "--copy bytes value must be a number zero or greater"
        ) "BYTES")
      "copy N bytes from input stream"
  , Option ['p'] ["outoffset"]
      (ReqArg (\p opts ->
        case reads p of
          [(outoffset, "")] | outoffset >= 0 -> Right opts { optOutOffset = outoffset }
          _ -> Left "--output offset value must be a number zero or greater"
        ) "OFFSET")
      "byte offset on output stream"
  , Option ['V','?'] ["version"]
        (NoArg (\ opts -> Right opts { optShowVersion = True }))  -- here no need for a value for this option because of `NoArg`
        "show version number"
  , Option ['v'] ["verbose"]
      (NoArg (\opts -> Right opts { optVerbose = True }))
      "verbose output"
  ]

parseArgs :: IO Options
parseArgs = do
  argv <- getArgs
  progName <- getProgName
  putStrLn progName
  -- putStrLn show (length progName)
  let header = "Usage: " ++ progName ++ " [OPTION...]"
  let helpMessage = usageInfo header options
  case getOpt RequireOrder options argv of
    (opts, [], []) ->
      case foldM (flip id) defaultOptions opts of
        Right opts -> return opts
        Left errorMessage -> ioError (userError (errorMessage ++ "\n" ++ helpMessage))
    (_, _, errs) -> ioError (userError (concat errs ++ helpMessage))


main :: IO ()
main = do

  putStrLn "Put binary portion from file, version 1.0"
  putStrLn "Copyright (C) 2025 Dmitry Stefankov. All Rights Reserved."

  options <- parseArgs
  putStrLn $ show options

  let infile  = optInputFilename options
  let outfile = optOutputFilename options
  putStrLn $ "outfile: " ++ outfile

  let inoffset1 = optInOffset options
  let inbytes1 = optInBytes options
  let in_big_ofs = toInteger inoffset1 :: Integer

  let outoffset1 = optOutOffset options
  let out_big_ofs = toInteger outoffset1 :: Integer

  let b1 = null (dropWhile isSpace infile)
  let b2 = null (dropWhile isSpace outfile)
  if b1 == True
    then exitWith ExitSuccess else putStrLn $ "infile: " ++ infile
  if b2 == True
    then exitWith ExitSuccess else putStrLn $ "outfile: " ++ outfile

  ihandle <- openBinaryFile infile ReadMode
  hSeek ihandle AbsoluteSeek in_big_ofs
  inputBytes <- B.hGetContents ihandle
  hClose ihandle

  ohandle <- openBinaryFile outfile ReadWriteMode
  hSeek ohandle AbsoluteSeek out_big_ofs
  hPrint ohandle inputBytes
  hClose ohandle

  putStrLn "All done."
